store/account.go

```diff
@@ -9,65 +10,83 @@ import (
 )
 
 type accountStore struct {
-	db         *leveldb.DB
-	addressMap map[crypto.Address]*account.Account
-	total      int32
+	db        *leveldb.DB
+	addrCache *lru.Cache[crypto.Address, *account.Account]
+	total     int32
 }
 
 func accountKey(addr crypto.Address) []byte { return append(accountPrefix, addr.Bytes()...) }
 
-func newAccountStore(db *leveldb.DB) *accountStore {
+func newAccountStore(db *leveldb.DB, cacheSize int) *accountStore {
 	total := int32(0)
-	numberMap := make(map[int32]*account.Account)
-	addressMap := make(map[crypto.Address]*account.Account)
+	addrLruCache, err := lru.New[crypto.Address, *account.Account](cacheSize)
+	if err != nil {
+		logger.Panic("unable to create new instance of lru cache", "error", err)
+	}
+
 	r := util.BytesPrefix(accountPrefix)
 	iter := db.NewIterator(r, nil)
 	for iter.Next() {
-		key := iter.Key()
-		value := iter.Value()
-
-		acc, err := account.FromBytes(value)
-		if err != nil {
-			logger.Panic("unable to decode account", "error", err)
-		}
-
-		var addr crypto.Address
-		copy(addr[:], key[1:])
-
-		numberMap[acc.Number()] = acc
-		addressMap[addr] = acc
 		total++
 	}
 	iter.Release()
 
 	return &accountStore{
-		db:         db,
-		total:      total,
-		addressMap: addressMap,
+		db:        db,
+		total:     total,
+		addrCache: addrLruCache,
 	}
 }
 
 func (as *accountStore) hasAccount(addr crypto.Address) bool {
-	_, ok := as.addressMap[addr]
+	ok := as.addrCache.Contains(addr)
+	if !ok {
+		ok = tryHas(as.db, accountKey(addr))
+	}
 	return ok
 }
 
 func (as *accountStore) account(addr crypto.Address) (*account.Account, error) {
-	acc, ok := as.addressMap[addr]
+	acc, ok := as.addrCache.Get(addr)
 	if ok {
 		return acc.Clone(), nil
 	}
 
-	return nil, ErrNotFound
+	rawData, err := tryGet(as.db, accountKey(addr))
+	if err != nil {
+		return nil, err
+	}
+
+	acc, err = account.FromBytes(rawData)
+	if err != nil {
+		return nil, err
+	}
+
+	as.addrCache.Add(addr, acc)
+	return acc, nil
```

Comment:
Here, I still feel that we need to review it again. We are returning an account pointer here, and it can be changed, which could potentially manipulate the cache.